### ISSUE TITLE
Rule Change: Add option to allow empty while loops to no-empty

### ISSUE THREAD

**Issue by Bwallker** [eslint/eslint#15839]
### What rule do you want to change?

no-empty

### What change to do you want to make?

Implement suggestions

### How do you think the change should be implemented?

A new option

### Example code

```js
const foo = () => {
  console.log('Still going');
  return Math.random() > 0.5;
};

while(foo()) {}
```

### What does the rule currently do for this code?

It generates an error because empty while loops are banned.

### What will the rule do after it's changed?

Generate an error based on if the allow-empy-while-loops option is enabled.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

This is useful when using Prettier because Prettier always formats empy while loops into empty blocks, which makes using Eslint and Prettier together really annoying because they can't agree on what to do with empty while loops. Currently the only way to solve this is by turning of no-empty, which I think is unfortunate because it's useful for catching other errors.

**Comment by mdjermanovic** [comment]
Hi @Bwallker, thanks for the issue!

Can you add a comment to the empty block? This rule allows empty blocks with comments inside as a way to mark blocks that are intentionally empty.

```js
const foo = () => {
  console.log('Still going');
  return Math.random() > 0.5;
};

while(foo()) { /* do nothing */ }
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWVtcHR5OiBlcnJvciAqL1xuXG5jb25zdCBmb28gPSAoKSA9PiB7XG4gIGNvbnNvbGUubG9nKCdTdGlsbCBnb2luZycpO1xuICByZXR1cm4gTWF0aC5yYW5kb20oKSA+IDAuNTtcbn07XG5cbndoaWxlKGZvbygpKSB7IC8qIGRvIG5vdGhpbmcgKi8gfVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

I'm not sure how common empty while loops are. It seems to me that it is better to mark a few that are intentionally empty than to allow all empty while loops and then miss the warnings for those that weren't intended to be empty.

**Comment by Bwallker** [comment]
> Hi @Bwallker, thanks for the issue!
> 
> Can you add a comment to the empty block? This rule allows empty blocks with comments inside as a way to mark blocks that are intentionally empty.
> 
> ```js
> const foo = () => {
>   console.log('Still going');
>   return Math.random() > 0.5;
> };
> 
> while(foo()) { /* do nothing */ }
> ```
> 
> [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWVtcHR5OiBlcnJvciAqL1xuXG5jb25zdCBmb28gPSAoKSA9PiB7XG4gIGNvbnNvbGUubG9nKCdTdGlsbCBnb2luZycpO1xuICByZXR1cm4gTWF0aC5yYW5kb20oKSA+IDAuNTtcbn07XG5cbndoaWxlKGZvbygpKSB7IC8qIGRvIG5vdGhpbmcgKi8gfVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)



Hello, thank you for the reply!

So, when I write empty while loops (which admitedly isn't every day, but it does happen), I will typically write them like this:

`while(foo());`

Then when I run Prettier over this it gets formatted to:

`while(foo()) {}`

Which Elint then complains about.

Now yes, I could manually go back and add a comment so Eslint stops complaining, but I find this disruptive to my work flow.

I realise that in the grand scale of things this is a fairly minor issue, but it still annoys me.

> 
> I'm not sure how common empty while loops are. It seems to me that it is better to mark a few that are intentionally empty than to allow all empty while loops and then miss the warnings for those that weren't intended to be empty.

You might be right. On further consideration it might be better to add a suggested fix that suggests adding a comment so it isn't as disruptive to work flow.

**Comment by amareshsm** [comment]
I am interested to work on this. can I pick this one?
@Bwallker let me know if you are working on this.

**Comment by Bwallker** [comment]
> I am interested to work on this. can I pick this one?
> @Bwallker let me know if you are working on this.

I'm not, just saw the reply. Go ahead please 👍

**Comment by mdjermanovic** [comment]
@amareshsm we only accept PRs for issues labeled "accepted", and this issue isn't accepted yet. You can find more details here:

https://eslint.org/docs/developer-guide/contributing/working-on-issues

**Comment by amareshsm** [comment]
@mdjermanovic  Makes sense to me, can you open a new change proposal with details about the suggested fix?  >> Change proposal means? Should I need a issue with the title as change request?

**Comment by mdjermanovic** [comment]
> Makes sense to me, can you open a new change proposal with details about the suggested fix? >> Change proposal means? Should I need a issue with the title as change request?

I suggested opening a separate issue for the functionality that was mentioned at the end of https://github.com/eslint/eslint/issues/15839#issuecomment-1119300679, because it is different from the original proposal.

The original proposal (this issue) is to add an option that would allow empty `while` loops, while the proposal in https://github.com/eslint/eslint/issues/15839#issuecomment-1119300679 is to add [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) to this rule.

**Comment by mdjermanovic** [comment]
> Got it. Where we can see these suggestions? It is mentioned suggestions will be shown in the editor but I am not able to reproduce them. Can someone help me with this? I would like to see how suggestions will be shown to the users.

VS Code supports suggestions.

Rules marked with 💡 in the [list of rules](https://eslint.org/docs/rules/) provide suggestions.

For example, if you have this code:

```js
/* eslint no-useless-escape: "error" */

"\a"
```

`\` will be underlined as an error. When you hover over it, VS Code will offer Quick Fix:

![image](https://user-images.githubusercontent.com/44349756/169520307-785c32c9-b25d-4525-a96e-ecfe562310e9.png)

![image](https://user-images.githubusercontent.com/44349756/169520833-4168a5bf-f770-4e33-b5e0-a584eb59e276.png)

First two options (`Remove the '\'...`, and `Replace the '\' with '\\'...`) are suggestions provided by the no-useless-escape rule. When you click on a suggestion, it will fix the code.

**Comment by mdjermanovic** [comment]
> > I'm not sure how common empty while loops are. It seems to me that it is better to mark a few that are intentionally empty than to allow all empty while loops and then miss the warnings for those that weren't intended to be empty.
> 
> You might be right. On further consideration it might be better to add a suggested fix that suggests adding a comment so it isn't as disruptive to work flow.

I agree. Since https://github.com/eslint/eslint/issues/15899 has been accepted, I'm closing this issue in favor of https://github.com/eslint/eslint/issues/15899.
